### ISSUE TITLE
Questions and ideas

### ISSUE THREAD

**Issue by yutian-zhao** [yutian-zhao/deep-learning-for-automated-planning#2]
# Model
1. Why can the model generalize to unseen problems or domains? 
If the data is representative to the problem set, then it makes sense. But in our case, small training problems are not necessarily representative for large problems. One possible reason mentioned in William's honors thesis is that Hypergraph Networks supports [combinatorial generalization](https://arxiv.org/abs/1903.12354) ([p32](https://shen.nz/thesis/shen-honours-19.pdf)). And moreover, we use optimal paths in the solutions as data, which may also not representative to the state distribution of the problem's state space.

2. How did William perform experiments (tune parameters)? It seems the performance varies a lot, how to get unbiased results efficiently?
How much effort was put into choosing what feature to use, getting the model to work and getting the training parameter?

3. Modify the model structure:
- [ ] Integrate constraints into the model, such as consistency, emphasizing the order/tendency and enlarging differences between values. (Tendency here refers to general ranking of states, and order here specifically refers to the order of the successors.)
- [ ] Modify encoding the features. Perhaps the order of the features also needs to be changed.
- [ ] Other ideas (low priority): Improve the linear regression module, use decay factor in the loss function, etc.
- [x] (High priority) Examine the model, especially the module coding part, whose implementation is not efficient and contains numerical issue.
- [ ] Issues are not clearly observed. May ask William's idea.

# Training
1. Given that we only pass the solutions, how does the model learn about unseen states? And how does it perform on unseen states (note here that we need to define unseen states (novel))?
- [x] (High priority) Check the convergence on training set and validation set, or even part of the test set since the validation set does not cover larger problems.

2. How does the model converges?
- [ ] Training parameters might need to be tuned for the model to be trained on more data.

What training method is helpful to train a general model? 
- [ ] Use regulation to avoid overfitting (original model uses weight decay in Adam to perform L2 regularization) and smooth the tendency. 

3. Is there any training technique to emphasize the more on the order of the predicted value. 
- [ ] Modify the loss function, like accompanying MSE with another loss function (need to be differentiable) [link](https://www.quora.com/What-do-you-suggest-as-a-loss-function-for-the-order-of-a-sequence), or penalizing more on states far away from the goals. 

4. If the order of loaded data has a certain pattern, will this make the model easier to learn?
- [x] (High priority) Examine the actual data loaded for training.

# Data 
1. What data is useful (i.e., not noise)? And especially, what data can help the model generalize to solve unseen problems?
- [ ] The key steps to take in order to solve a specific domain, which revealing the tricks.
- [x] (High priority) Novel states, especially unseen proposition combination. This can help the model distinguish bad states which are usually unseen in an optimal path from good states. (Does the definition of novelty work as we expected?) (Think about and try out different ways to deal with novelty.)

2. Why using all of the optimal paths between any two states leads to worse performance? 
- [ ] The data is noisy since it contains states which are not similar to goals (see (3)).
- [x] The data distribution is unbalanced. However, will random sampling ruin some property of the dataset? ~~We can resample from the dataset to match the distribution of the original dataset (but the original dataset is also not balanced).~~

3. How to determine states which have similar structure as goals?
- [ ] (Abandoned) Constraining the propositions of states to match the propositions of goals, i.e., have the same predicates (or no less than goals?) and have the same number of propositions for each proposition.
